flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -0,0 +1,245 @@
+        }
+
+        private Optional<EventHandler> findHandler(T obj) {
+            final String resourceKey = getResourceKey(obj.getMetadata().getName());
+            return Optional.ofNullable(handlers.get(resourceKey));
+        }
+    }
+
+    private class EventHandler {
+        private final String resourceKey;
+        private final Map<String, WatchCallbackHandler<R>> handlers = new HashMap<>();
+
+        private T currentResource;
+
+        private EventHandler(String resourceKey) {
+            this.resourceKey = resourceKey;
+        }
+
+        private void addWatch(String id, WatchCallbackHandler<R> handler) {
+            log.info("Starting to watch for {}, watching id:{}", resourceKey, id);
+            handlers.put(id, handler);
+            if (currentResource != null) {
+                final List<R> events =
+                        Collections.singletonList(eventWrapper.apply(currentResource));
+                safelyInvokeHandler(id, handler::onAdded, events);
+            }
+        }
+
+        private boolean removeWatch(String id) {
+            handlers.remove(id);
+            log.info("Stopped to watch for {}, watching id:{}", resourceKey, id);
+            return handlers.isEmpty();
+        }
+
+        private void handleResourceEvent() {
+            T newResource = sharedIndexInformer.getIndexer().getByKey(resourceKey);
+            T oldResource = this.currentResource;
+            if (newResource == null) {
+                if (oldResource != null) {
+                    List<R> events = Collections.singletonList(eventWrapper.apply(oldResource));
+                    this.handlers.forEach(
+                            (id, handler) -> safelyInvokeHandler(id, handler::onDeleted, events));
+                }
+            } else {
+                List<R> events = Collections.singletonList(eventWrapper.apply(newResource));
+                if (oldResource == null) {
+                    this.handlers.forEach(
+                            (id, handler) -> safelyInvokeHandler(id, handler::onAdded, events));
+                } else if (!oldResource
+                        .getMetadata()
+                        .getResourceVersion()
+                        .equals(newResource.getMetadata().getResourceVersion())) {
+
+                    this.handlers.forEach(
+                            (id, handler) -> safelyInvokeHandler(id, handler::onModified, events));
+                }
+            }
+            this.currentResource = newResource;
+        }
+
+        private void handleExceptionEvent(Exception e) {
+            this.handlers.forEach(
+                    (id, handler) -> safelyInvokeHandler(id, handler::handleError, e));
+        }
+
+        private <E> void safelyInvokeHandler(String id, Consumer<E> eventHandler, E event) {
+            try {
+                eventHandler.accept(event);
+            } catch (Exception e) {
+                log.error(
```

Comment:
I think we need some proper error handling here. If anything goes wrong, we can get into an undefined state (eg. not knowing about leader, ...).

Reply:
Probably simply not catching the exception should be sufficient, as it will propagate all the way to FatalExitExceptionHandler (default in ExecutorThreadFactory)